### DISCUSSION TITLE
Signature Generation using CryptoJS

### DISCUSSION THREAD

**Discussion by dirkkruger** [minio/minio#15015] [Q&A]
Hi all

I have a requirement to retrieve objects from MINIO Storage, this works fine as long as the access type is set to Public by simply accessing the URL, but the organization wishes to set it to Private, hence the requirement for the generation of a signature.

The environment I am working in is ServiceNow. As part of the solution, I have added the CryptoJS library to ServiceNow to assist in generation of my signature, but my result suggests my Signature does not match the calculated signature.

I've used the following references as a guide:
https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
https://docs.aws.amazon.com/general/latest/gr/signature-v4-examples.html#signature-v4-examples-javascript

Can somebody please assist? Below is the code I am using as an experiment.

```
var sigDate = new GlideDateTime();
var dateStamp = String(sigDate).substr(0, 4) + String(sigDate).substr(5, 2) + String(sigDate).substr(8, 2);
var timeStamp = String(sigDate).substr(11, 2) + String(sigDate).substr(14, 2) + String(sigDate).substr(17, 2);

var CanonicalRequest;
var httpVerb = "GET";
var canonicalURI = "/Link.png";
var canonicalQueryString = "X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=" + <AccessKey>+ "/" + dateStamp + "/me-south-1/s3/aws4_request&X-Amz-Date=" + dateStamp + "T" + timeStamp + "Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host";
var host = "host:objectapi-stg.xxx.xxx.xxx";
CanonicalRequest = httpVerb + "\n" + canonicalURI + "\n" + canonicalQueryString + "\n" + host + "\n" + "host\nUNSIGNED-PAYLOAD";

var StringToSign;
var AWS = "AWS4-HMAC-SHA256";
var STStimeStamp = dateStamp + "T" + timeStamp + "Z";
var STSscope = dateStamp + "/me-south-1/s3/aws4_request";
var hexString = CryptoJS.enc.Hex.stringify(CryptoJS.SHA256(CanonicalRequest));

StringToSign = AWS + "\n" + STStimeStamp + "\n" + STSscope + "\n" + hexString;

var Signature;
var SignatureKey = getSignatureKey(<SecretKey>), dateStamp, "me-south-1, "s3"));

Signature = CryptoJS.enc.Hex.stringify(CryptoJS.HmacSHA256(SignatureKey, StringToSign));

gs.info("https://objectapi-stg.xxx.xxx.xxx" + canonicalURI + "?" + canonicalQueryString + "&X-Amz-Signature=" + Signature);

function getSignatureKey(key, dateStamp, regionName, serviceName) {
    var kDate = CryptoJS.HmacSHA256(dateStamp, "AWS4" + key);
    var kRegion = CryptoJS.HmacSHA256(regionName, kDate);
    var kService = CryptoJS.HmacSHA256(serviceName, kRegion);
    var kSigning = CryptoJS.HmacSHA256("aws4_request", kService);
    return kSigning;
}
```
The output I received from the above script is as follows:
**https://objectapi-stg.xxx.xxx.xxx/Link.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AccessKey/20220601/me-south-1/s3/aws4_request&X-Amz-Date=20220601T104918Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=1ca9e4f80fb34b3ed8ddb742a28874d2800894b67e4054c91612ca3c07a9237a**

And the error I received is as follows:
```
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>
The request signature we calculated does not match the signature you provided. Check your key and signing method.
</Message>
<Key>Link.png</Key>
<BucketName>xxx</BucketName>
<Resource>/Link.png</Resource>
<Region>me-south-1</Region>
<RequestId>16F477BD85CA0B34</RequestId>
<HostId>4a75b69c-71e5-4c9f-9da2-fc9e459e75c0</HostId>
</Error>
```

**Comment by harshavardhana** [comment]
Your canonical request is wrong read the AWS S3 signature docs on how to do this properly.

**Reply by harshavardhana** [reply]
why are you wasting time writing your own signature v4 implementation  - why can't use you that is already available..  read on how to do this properly https://github.com/minio/minio-js/blob/master/src/main/signing.js
